### PR TITLE
Coercion strategy hierarchies [PoC]

### PR DESCRIPTION
**Pull request by camsaul**
Adds dynamically generated keyword hierarchies for coercion strategies allowed base type(s) and resulting effective type, meaning we don't have to duplicate all this logic ourselves and defining new strategies can be done declaratively.

I put the tooling behind this in a separate namespace so we don't have to look at it every time we poke around `types.cljc`

I didn't port the old stuff over yet -- waiting for feedback first

### CODE REVIEW

File: shared/.dir-locals.el [hunk 1]
```diff
@@ -1 +1,3 @@
-((clojure-mode . ((cider-preferred-build-tool . shadow-cljs))))
+((nil . ((cider-default-cljs-repl . shadow-select)
+         (cider-shadow-default-options . "node-repl")))
+ (clojure-mode . ((cider-preferred-build-tool . shadow-cljs))))
```
Comment (dpsutton) [inline shared/.dir-locals.el:3]: you might want nil or clojurec-mode here. In the clojurec buffers i suspect you won't get the preferred-build-tool option

Comment (camsaul) [reply shared/.dir-locals.el:3]: `clojure-mode` is the parent mode of `clojurec-mode`, so it actually still works in that case. It's also the parent mode for `clojurescript-mode` so I guess these can all go under `clojure-mode`

Comment (dpsutton) [reply shared/.dir-locals.el:3]: ok. i had forgotten if it inherited like that. i haven't used CIDER in a while so it wouldn't have impacted me.

**Review by camsaul — COMMENTED** [review]
@dpsutton your changes look good to me 👍

**Review by dpsutton — APPROVED** [review]

**Pull request merged**